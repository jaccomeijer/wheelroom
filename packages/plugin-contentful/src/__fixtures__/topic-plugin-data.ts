import { PushHandlerArgs } from '@wheelroom/push/plain'

export const topicPluginData: PushHandlerArgs = {
  callCommand: 'push',
  callType: 'models',
  pluginData: {
    dataVar: {
      fileName:
        '/Users/jacco/Development/wheelroom/packages/push/src/lib/get-all-plugin-data/__fixtures__/topic.ts',
      isArray: false,
      isExported: true,
      isObject: true,
      isTyped: true,
      name: 'wheelroomPluginData',
      type: 'WheelroomPluginData',
      value: `{
  '@wheelroom/plugin-contentful/plain': {
    validations: {
      dropdown: { in: ['item A', 'item B', 'item C'] },
      media: { linkMimetypeGroup: ['image', 'video'] },
      action: { linkContentType: ['action'] },
      shortString: { size: { max: 155, min: 0 } },
      camelCase: {
        regexp: { pattern: '^[a-z]+([A-Z][a-z0-9]+)*$', flags: '' },
      },
      path: {
        regexp: {
          pattern: '^\\\\\\\\/[a-z0-9:\\\\\\\\._/~%\\\\\\\\-\\\\\\\\+&\\\\\\\\#\\\\\\\\?!=\\\\\\\\(\\\\\\\\)@]*$',
          flags: '',
        },
      },
      url: {
        regexp: {
          pattern:
            '^(ftp|http|https|mailto):(\\\\\\\\/\\\\\\\\/)*(\\\\\\\\w+:{0,1}\\\\\\\\w*@)?(\\\\\\\\S+)(:[0-9]+)?(\\\\\\\\/|\\\\\\\\/([\\\\\\\\w#!:.?+=&%@!\\\\\\\\-/]))?$',
          flags: '',
        },
      },
    },
  },
}`,
    },
    types: {
      ContentfulAction: {
        interface: {
          fieldTags: {
            anchor: {
              '@helpText': 'Voorbeeld: mijn-anchor',
              '@type': 'Symbol',
            },
            description: {
              '@helpText': 'Omschrijving van de actie, voor zoekmachines',
              '@type': 'Symbol',
            },
            eventId: {
              '@type': 'Symbol',
            },
            heading: {
              '@type': 'Symbol',
            },
            query: {
              '@type': 'Symbol',
            },
            url: {
              '@type': 'Symbol',
            },
          },
          interfaceTags: {
            '@plugin': '@wheelroom/plugin-contentful/plain',
            '@type': 'testAction',
          },
          typeName: 'ContentfulAction',
        },
        variables: [],
      },
      ContentfulTopic: {
        interface: {
          fieldTags: {
            abstract: {
              '@type': 'Symbol',
              '@validation': 'shortString',
            },
            actionsCollection: {
              '@description': 'Action field, this is the description',
              '@itemsLinkType': 'Entry',
              '@itemsType': 'Link',
              '@type': 'Array',
              '@validation': 'action',
            },
            heading: {
              '@type': 'Symbol',
            },
            icon: {
              '@type': 'Symbol',
            },
            mediaCollection: {
              '@itemsLinkType': 'Asset',
              '@itemsType': 'Link',
              '@type': 'Array',
            },
            poster: {
              '@linkType': 'Asset',
              '@type': 'Link',
            },
          },
          interfaceTags: {
            '@description':
              'Topic model, a heading, an abstract and a call to action',
            '@displayField': 'heading',
            '@plugin': '@wheelroom/plugin-contentful/plain',
            '@type': 'testTopic',
          },
          typeName: 'ContentfulTopic',
        },
        variables: [
          {
            fileName:
              '/Users/jacco/Development/wheelroom/packages/push/src/lib/get-all-plugin-data/__fixtures__/topic.ts',
            isArray: true,
            isExported: true,
            isObject: false,
            isTyped: true,
            name: 'myTopics',
            type: 'ContentfulTopic',
            value: `[
  {
    abstract: 'This is the abstract',
    actionsCollection: {
      items: [
        {
          heading: 'Hello',
          anchor: 'anchor',
        },
      ],
    },
    heading: 'Heres the heading',
  },
]`,
          },
        ],
      },
    },
  },
}
