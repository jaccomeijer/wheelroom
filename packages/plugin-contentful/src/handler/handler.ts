import chalk from 'chalk'
import { PushHandler } from '@wheelroom/push/plain'
import { getWheelroomPluginData } from '../lib/get-wheelroom-plugin-data'
import { getAndValidateEnv } from '../lib/get-and-validate-env'
import { pushModels } from '../push-models/push-models'
import { pushContent } from '../push-content/push-content'
import { getContentfulEnvironment } from '../lib/get-contentful-environment'

export const handler: PushHandler = async ({
  callCommand,
  callType,
  pluginData,
}) => {
  const log = console.log
  getAndValidateEnv()
  log(chalk.bold.underline(`Starting Contentful plugin`))
  if (!pluginData) return
  log(`Wheelroom project: ${process.env.WHEELROOM_PROJECT_NAME}`)

  const dataVarObj = getWheelroomPluginData({ dataVar: pluginData.dataVar })
  const validationsMap =
    dataVarObj?.['@wheelroom/plugin-contentful/plain']?.validations || {}
  const contentfulEnvironment = await getContentfulEnvironment()
  if (callCommand === 'push' && callType === 'models') {
    await pushModels({
      contentfulEnvironment,
      typeData: pluginData.types,
      validationsMap,
    })
  }
  if (callCommand === 'push' && callType === 'content') {
    await pushContent({
      contentfulEnvironment,
      typeData: pluginData.types,
    })
  }
}
