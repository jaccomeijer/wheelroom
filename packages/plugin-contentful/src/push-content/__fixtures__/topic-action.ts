import { TypeData } from '@wheelroom/push/plain'

export const topicAction: TypeData = {
  ContentfulAction: {
    interface: {
      fieldTags: {
        heading: {
          '@type': 'Symbol',
        },
      },
      interfaceTags: {
        '@displayField': 'heading',
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'testAction',
      },
      typeName: 'ContentfulAction',
    },
    variables: [
      {
        fileName: '/__fixtures__/action.ts',
        name: 'myActions',
        type: 'ContentfulAction',
        value: `[
  {
    heading: 'My action heading #0',
  },
]`,
      },
    ],
  },
  ContentfulTopic: {
    interface: {
      fieldTags: {
        heading: {
          '@type': 'Symbol',
        },
        actionsCollection: {
          '@description': 'Action field, this is the description',
          '@itemsLinkType': 'Entry',
          '@itemsType': 'Link',
          '@type': 'Array',
          '@validation': 'action',
        },
      },
      interfaceTags: {
        '@displayField': 'heading',
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'testTopic',
      },
      typeName: 'ContentfulTopic',
    },
    variables: [
      {
        fileName: '/__fixtures__/topic.ts',
        name: 'myTopics',
        type: 'ContentfulTopic',
        value: `[
  {
    heading: 'My topic heading #0',
    actionsCollection: {
      items: [
        {
          heading: 'Action inline',
        },
        {
          sys: {
            id: 'myActions0',
          }
        },
      ],
    },
  },
]`,
      },
    ],
  },
}
