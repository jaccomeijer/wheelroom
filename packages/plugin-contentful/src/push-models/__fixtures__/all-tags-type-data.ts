import { TypeData } from '@wheelroom/push/plain'

export const allTagsTypeData: TypeData = {
  Topic: {
    interface: {
      fieldTags: {
        heading: {
          '@helpText': 'The topic heading',
          '@itemsLinkType': 'Entry',
          '@itemsType': 'Link',
          '@linkType': 'Symbol',
          '@localized': '',
          '@name': 'Heading',
          '@required': '',
          '@type': 'Array',
          '@validation': 'unique',
          '@widget': 'singleLine',
        },
      },
      interfaceTags: {
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'topic',
      },
      typeName: 'Topic',
    },
    variables: [],
  },
}
