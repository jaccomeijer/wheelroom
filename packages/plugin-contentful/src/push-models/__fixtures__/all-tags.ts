import { TypeData } from '@wheelroom/push/plain'

export const allTags: TypeData = {
  Topic: {
    interface: {
      fieldTags: {
        heading: {
          '@helpText': 'The topic heading',
          '@itemsLinkType': 'Entry',
          '@itemsType': 'Link',
          '@linkType': 'Symbol',
          '@localized': '',
          '@name': 'Heading',
          '@required': '',
          '@type': 'Array',
          '@validation': 'mockedValidation',
          '@widget': 'singleLine',
        },
      },
      interfaceTags: {
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'topic',
      },
      typeName: 'Topic',
    },
    variables: [],
  },
}
