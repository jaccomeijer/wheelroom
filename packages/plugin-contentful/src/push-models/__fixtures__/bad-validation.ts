import { TypeData } from '@wheelroom/push/plain'

export const badValidation: TypeData = {
  Topic: {
    interface: {
      fieldTags: {
        heading: {
          '@type': 'Symbol',
          '@validation': 'doesNotExist',
        },
      },
      interfaceTags: {
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'topic',
      },
      typeName: 'Topic',
    },
    variables: [],
  },
}
