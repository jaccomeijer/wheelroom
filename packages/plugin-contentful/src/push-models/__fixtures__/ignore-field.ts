import { TypeData } from '@wheelroom/push/plain'

export const ignoreField: TypeData = {
  Topic: {
    interface: {
      fields: {
        heading: {
          tags: {
            '@helpText': 'The topic heading',
            '@ignore': '',
            '@type': 'Symbol',
          },
          type: 'string | undefined',
        },
      },
      interfaceTags: {
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'topic',
      },
      typeName: 'Topic',
    },
    variables: [],
  },
}
