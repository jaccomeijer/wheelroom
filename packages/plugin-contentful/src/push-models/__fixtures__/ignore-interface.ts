import { TypeData } from '@wheelroom/push/plain'

export const ignoreInterface: TypeData = {
  Topic: {
    interface: {
      fields: {
        heading: {
          tags: {
            '@helpText': 'The topic heading',
            '@type': 'Symbol',
          },
          type: 'string | undefined',
        },
      },
      interfaceTags: {
        '@ignore': '',
        '@plugin': '@wheelroom/plugin-contentful/plain',
        '@type': 'topic',
      },
      typeName: 'Topic',
    },
    variables: [],
  },
}
