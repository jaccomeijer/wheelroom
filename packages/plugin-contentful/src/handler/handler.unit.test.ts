const contentTypeMock = {
  publish: async () => contentTypeMock,
  update: async () => contentTypeMock,
  getEditorInterface: async () => contentTypeMock,
}
const getContentType = jest.fn(async () => contentTypeMock)
const createContentTypeWithId = jest.fn(async () => contentTypeMock)
const getEnvironment = jest.fn(async () => ({
  createContentTypeWithId,
  getContentType,
}))
const getSpace = jest.fn(async () => ({ getEnvironment }))
const createClient = jest.fn(() => ({ getSpace }))
jest.mock('contentful-management', () => ({ createClient }))

import { handler } from './handler'
import { topicPluginData } from './__fixtures__/topic-plugin-data'

describe('The handler should', () => {
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation()
  beforeEach(() => {
    consoleSpy.mockReset()
  })

  test('run with topic data', async () => {
    await handler(topicPluginData)
    expect(consoleSpy.mock.calls).toMatchSnapshot()
  })
})
