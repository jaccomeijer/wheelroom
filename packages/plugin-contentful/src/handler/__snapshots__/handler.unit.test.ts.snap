// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The handler should run with topic data 1`] = `
Array [
  Array [
    "============ Starting Contentful plugin",
  ],
  Array [
    "Wheelroom project: undefined",
  ],
  Array [
    "Processing: ContentfulAction...",
  ],
  Array [
    "Adding editor control for field anchor",
  ],
  Array [
    "Adding editor control for field description",
  ],
  Array [
    "Contentful API, updating existing testAction",
  ],
  Array [
    "Contentful API, updating editor interface",
  ],
  Array [
    "Processing: ContentfulTopic...",
  ],
  Array [
    "Adding validation shortString",
  ],
  Array [
    "Adding validation action",
  ],
  Array [
    "Contentful API, updating existing testTopic",
  ],
  Array [
    "Contentful API, updating editor interface",
  ],
  Array [
    "Data to send to Contentful",
    "[
  {
    abstract: [32m'This is the abstract'[39m,
    actionsCollection: { items: [ { heading: [32m'Hello'[39m, anchor: [32m'anchor'[39m } ] },
    heading: [32m'Heres the heading'[39m
  }
]",
  ],
]
`;
